packages/kbn-config-schema/index.ts

```diff
@@ -199,6 +202,164 @@ function oneOf<RTS extends Array<Type<any>>>(
   return new UnionType(types, options);
 }
 
+function allOf<
+  A extends Props,
+  B extends Props,
+  C extends Props,
+  D extends Props,
+  E extends Props,
+  F extends Props,
+  G extends Props,
+  H extends Props,
+  I extends Props,
```

Comment:
So, this is absolutely atrocious, and I swear I spent more time trying to have a "smart inference" based on the array of received types rather than going down the multi-ultraverbose-signature path, but in the end, it's the only approach, as there's no way for typescript to properly infer the generic subtype of each individual element in a array of generics (e.g. `Array<Type<something extends T>>` -> T can't be properly inferred per element in the array).

(but this is already was we were doing of `anyOf`, fwiw)

Reply:
++ I don't think we have to solve this now, but if we do it would be good to also do so for `oneOf`.

In a draft PR I managed to simplify `oneOf`'s type declration. It's a slight refactor for any users of the generic types (i.e. users that are passing in types via `oneOf<string, number...>`. It boils down to making `RTS` an inferred variadic array:

```ts
function oneOf<RTS extends Array<Type<unknown>>>(
  types: [...RTS],
  options?: UnionTypeOptions<any>
): Type<TypeOf<RTS[number]>> {
  return new UnionType(types, options);
}
```

Then:

```
const b = oneOf([string(), number()]);
=> Type<string | number>
```

This is sort of a happy accident of how `RTS[number]` combines types using `|`. But perhaps it is possible to go from `A | B | C` => `A & B & C` 🤷🏻 .